From 7e0f5149c47f6b3934c5b64626068fdd7dabb0a8 Mon Sep 17 00:00:00 2001
From: Vincent Rabaud <vincent.rabaud@gmail.com>
Date: Sun, 13 Mar 2016 18:05:53 +0100
Subject: [PATCH 7/8] add C++11 flags to whoever uses that lib

---
 cmake/libqi-extras.cmake.em | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/cmake/libqi-extras.cmake.em b/cmake/libqi-extras.cmake.em
index a720060..894d2ed 100644
--- a/cmake/libqi-extras.cmake.em
+++ b/cmake/libqi-extras.cmake.em
@@ -1,3 +1,6 @@
 @[if DEVELSPACE]@
 include_directories(@(CATKIN_DEVEL_PREFIX)/include)
 @[end if]@
+if ("${CMAKE_CXX_COMPILER_ID}" STREQUAL "Clang" OR "${CMAKE_CXX_COMPILER_ID}" STREQUAL "GNU")
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -std=gnu++11")
+endif()
-- 
1.9.1

